src/nonFree/rich_sift/Rich_SIFT_describer.hpp

```diff
@@ -0,0 +1,321 @@
+    return true;
+  }
+
+  /**
+  @brief Detect regions on the image and compute their attributes (description)
+  @param image Image.
+  @param regions The detected regions and attributes
+  @param mask 8-bit gray image for keypoint filtering (optional).
+     Non-zero values depict the region of interest.
+  */
+  bool Describe
+  (
+    const image::Image<unsigned char> & image,
+    std::unique_ptr<Regions> &regions,
+    const image::Image<unsigned char> * mask = nullptr
+  ) override
+  {
+    Regions_type *regions_type = static_cast<Regions_type*>(regions.get());
+    if (regions_type == nullptr || regions_type->RegionCount() == 0)
+    {
+      regions = DescribeSIFT(image, mask);
+    }
+    else
+    {
+      DescribeSIFTGivenRegion(image, regions, mask);
+    }
+    return regions != nullptr;
+  }
+
+  /**
+  @brief Detect regions on the image and compute their attributes (description)
+  @param image Image.
+  @param mask 8-bit gray image for keypoint filtering (optional).
+     Non-zero values depict the region of interest.
+  @return regions The detected regions and attributes (the caller must delete the allocated data)
+  */
+  std::unique_ptr<Regions> Describe(
+    const image::Image<unsigned char>& image,
+    const image::Image<unsigned char>* mask = nullptr
+  ) override
+  {
+    return DescribeSIFT(image, mask);
+  }
+
+  /**
+  @brief compute given regions' attributes (description)
+  @param image Image.
+  @param mask 8-bit gray image for keypoint filtering (optional).
+     Non-zero values depict the region of interest.
+  (the caller must delete the allocated data)
+  */
+  void DescribeSIFTGivenRegion(
+      const image::Image<unsigned char> &image,
+      std::unique_ptr<Regions> &regions,
+      const image::Image<unsigned char>* mask = nullptr
+  )
+  {
+    Regions_type *regions_type = static_cast<Regions_type*>(regions.get());
+
+    const int w = image.Width(), h = image.Height();
+    //Convert to float
+    const image::Image<float> If(image.GetMat().cast<float>());
+
+    VlSiftFilt *filt = vl_sift_new(w, h,
+      _params._num_octaves, _params._num_scales, _params._first_octave);
+    if (_params._edge_threshold >= 0)
+      vl_sift_set_edge_thresh(filt, _params._edge_threshold);
+    if (_params._peak_threshold >= 0)
+      vl_sift_set_peak_thresh(filt, 255*_params._peak_threshold/_params._num_scales);
+
+    Descriptor<vl_sift_pix, 128> descr;
+    Descriptor<unsigned char, 128> descriptor;
+
+    // Process SIFT computation
+    vl_sift_process_first_octave(filt, If.data());
+
+    // reserve some memory for faster keypoint saving
+    auto& features = regions_type->Features();
+    regions_type->Descriptors().clear();
+    regions_type->Descriptors().reserve(2000);
+
+    int curr_feature_index = 0;
+    int octave_index = 0;
+    while (true) {
+      // Update gradient before launching parallel extraction
+      vl_sift_update_gradient(filt);
+
+      while (curr_feature_index < features.size() &&
```

Comment:
use openmp here to accelerate computation

Reply:
done